arangod/Aql/CollectNode.cpp

```diff
@@ -125,8 +126,73 @@ void CollectNode::toVelocyPackHelper(VPackBuilder& nodes, unsigned flags) const
+        writeableOutputRegisters.insert((*it).second.registerId);
+      }
+
+      std::vector<std::pair<RegisterId, RegisterId>> groupRegisters;
+      for (auto const& p : _groupVariables) {
+        // We know that planRegisters() has been run, so
+        // getPlanNode()->_registerPlan is set up
+        auto itOut = getRegisterPlan()->varInfo.find(p.first->id);
+        TRI_ASSERT(itOut != getRegisterPlan()->varInfo.end());
+
+        auto itIn = getRegisterPlan()->varInfo.find(p.second->id);
+        TRI_ASSERT(itIn != getRegisterPlan()->varInfo.end());
+        TRI_ASSERT((*itIn).second.registerId < ExecutionNode::MaxRegisterId);
+        TRI_ASSERT((*itOut).second.registerId < ExecutionNode::MaxRegisterId);
+        groupRegisters.emplace_back(std::make_pair((*itOut).second.registerId,
+                                                   (*itIn).second.registerId));
+        writeableOutputRegisters.insert((*itOut).second.registerId);
+        readableInputRegisters.insert((*itIn).second.registerId);
+      }
+
+      std::vector<std::pair<RegisterId, RegisterId>> aggregateRegisters;
+      for (auto const& p : _aggregateVariables) {
+        // We know that planRegisters() has been run, so
+        // getPlanNode()->_registerPlan is set up
+        auto itOut = getRegisterPlan()->varInfo.find(p.first->id);
+        TRI_ASSERT(itOut != getRegisterPlan()->varInfo.end());
+        TRI_ASSERT((*itOut).second.registerId < ExecutionNode::MaxRegisterId);
+
+        RegisterId reg;
+        if (Aggregator::requiresInput(p.second.second)) {
+          auto itIn = getRegisterPlan()->varInfo.find(p.second.first->id);
+          TRI_ASSERT(itIn != getRegisterPlan()->varInfo.end());
+          TRI_ASSERT((*itIn).second.registerId < ExecutionNode::MaxRegisterId);
+          reg = (*itIn).second.registerId;
+          readableInputRegisters.insert((*itIn).second.registerId);
+        } else {
+          // no input variable required
+          reg = ExecutionNode::MaxRegisterId;
+        }
+        aggregateRegisters.emplace_back(std::make_pair((*itOut).second.registerId, reg));
+        writeableOutputRegisters.insert((*itOut).second.registerId);
+      }
+      TRI_ASSERT(aggregateRegisters.size() == _aggregateVariables.size());
+
+      transaction::Methods* trxPtr = _plan->getAst()->query()->trx();
+      HashedCollectExecutorInfos infos(
+          getRegisterPlan()->nrRegs[previousNode->getDepth()],
+          getRegisterPlan()->nrRegs[getDepth()], getRegsToClear(), calcRegsToKeep(),
+          std::move(readableInputRegisters), std::move(writeableOutputRegisters),
+          std::move(groupRegisters), collectRegister, _aggregateVariables,
+          std::move(aggregateRegisters), _groupVariables, trxPtr, _count);
+
+      return std::make_unique<ExecutionBlockImpl<HashedCollectExecutor>>(&engine, this,
+                                                                         std::move(infos));
```

Comment:
@hkernbach Please extract some methods like `getCollectRegister()`, `calcGroupRegisters()`, `calcAggregateRegisters()` or so in order to preserve a clearer view to what happens here.